ext/limiter/limiter.c

```diff
@@ -0,0 +1,158 @@
+
+#include <pthread.h>
+#include <unistd.h>
+#include <sys/mman.h>
+
+// clang-format Off
+
+#define NANOSECONDS_PER_SECOND 1000000000
+
+typedef struct {
+    /* limit from configuration DD_TRACE_RATE_LIMIT */
+    uint32_t limit;
+    struct {
+        /* beginning of window */
+        uint64_t open;
+        /* last hit inside this window */
+        uint64_t current;
+        /* total count for this window */
+        uint32_t samples;
+        /* allowed count for this window */
+        uint32_t allowed;
+        /* limit (tokens) for this window */
+        uint32_t limit;
+        /* rate for the last window that passed */
+        uint32_t rate;
+    } window;
+
+    pthread_mutex_t lock;
+} ddtrace_limiter;
+
+static ddtrace_limiter* __limiter;
+
+static inline uint64_t ddtrace_limiter_clock() {
+    struct timespec ts;
+
+    if (clock_gettime(CLOCK_MONOTONIC, &ts) != 0) {
+        /* TODO something, can this fail ? */
+        return 0;
+    }
+
+    return (ts.tv_sec * NANOSECONDS_PER_SECOND) + /* seconds as nanoseconds */
+            ts.tv_nsec;                           /* plus remaining nanoseconds */
+}
+
+void ddtrace_limiter_create() {
+    uint32_t limit = (uint32_t) get_global_DD_TRACE_RATE_LIMIT();
+
+    if (!limit) {
+        return;
+    }
+
+    __limiter = mmap(NULL, sysconf(_SC_PAGESIZE), PROT_WRITE, MAP_SHARED|MAP_ANONYMOUS, -1, 0);
+
+    if ( __limiter == MAP_FAILED) {
+        __limiter = NULL;
+        return;
+    }
+
+    __limiter->limit = limit;
+    __limiter->window.open = ddtrace_limiter_clock();
+    __limiter->window.limit = limit;
+
+    pthread_mutexattr_t mattr;
+
+    pthread_mutexattr_init(&mattr);
+    pthread_mutexattr_setpshared(
+        &mattr, PTHREAD_PROCESS_SHARED);
+    pthread_mutex_init(&__limiter->lock, &mattr);
+    pthread_mutexattr_destroy(&mattr);
+}
+
+bool ddtrace_limiter_active() {
+    if (!__limiter) {
+        return false;
+    }
+
+    return true;
+}
+
+bool ddtrace_limiter_allow() {
+    ZEND_ASSERT(__limiter);
+
+    pthread_mutex_lock(&__limiter->lock);
+
+    uint64_t previous = __limiter->window.current;
+
+    __limiter->window.current  = ddtrace_limiter_clock();
+
+    uint64_t elapsed = __limiter->window.current - previous;
+
+    if (__limiter->window.current > (__limiter->window.open + NANOSECONDS_PER_SECOND)) {
+        /* window passed, reset */
+
+        /* store current window rate for effective rate calculation */
+        __limiter->window.rate = __limiter->window.samples ?
+            (double) __limiter->window.allowed / __limiter->limit : 1;
+
+        /* move window */
+        __limiter->window.open = __limiter->window.current;
+        __limiter->window.samples = 0;
+        __limiter->window.allowed = 0;
+    }
+
+    if (__limiter->window.limit != __limiter->limit) {
+        __limiter->window.limit = MIN(
+            __limiter->limit,
+            __limiter->window.limit + ((elapsed / NANOSECONDS_PER_SECOND) * __limiter->limit));
+    }
```

Comment:
When is that ever true? I don't see any assignments except at startup time?

Reply:
> When is that ever true?

Never :)